cmd/zpool/zpool_main.c

```diff
@@ -822,90 +828,107 @@ zpool_do_labelclear(int argc, char **argv)
 
 	/* get vdev name */
 	if (argc < 1) {
-		(void) fprintf(stderr, gettext("missing vdev device name\n"));
+		(void) fprintf(stderr, gettext("missing vdev name\n"));
 		usage(B_FALSE);
 	}
-
-	vdev = argv[0];
-	if ((fd = open(vdev, O_RDWR)) < 0) {
-		(void) fprintf(stderr, gettext("Unable to open %s\n"), vdev);
-		return (B_FALSE);
+	if (argc > 1) {
+		(void) fprintf(stderr, gettext("too many arguments\n"));
+		usage(B_FALSE);
 	}
 
-	name = NULL;
-	if (zpool_in_use(g_zfs, fd, &state, &name, &inuse) != 0) {
-		if (force)
-			goto wipe_label;
+	/*
+	 * Check if we were given absolute path and use it as is.
+	 * Otherwise if the provided vdev name doesn't point to a file,
+	 * try prepending expected disk paths and partition numbers.
+	 */
+	(void) strlcpy(vdev, argv[0], sizeof (vdev));
+	if (vdev[0] != '/' && stat(vdev, &st) != 0) {
+		int error;
+
+		error = zfs_resolve_shortname(argv[0], vdev, MAXPATHLEN);
+		if (error == 0 && zfs_dev_is_whole_disk(vdev)) {
+			if (zfs_append_partition(vdev, MAXPATHLEN) == -1)
+				error = ENOENT;
+		}
 
-		(void) fprintf(stderr,
-		    gettext("Unable to determine pool state for %s\n"
-		    "Use -f to force the clearing any label data\n"), vdev);
+		if (error || (stat(vdev, &st) != 0)) {
+			(void) fprintf(stderr, gettext(
+			    "failed to find device %s, try specifying absolute "
+			    "path instead\n"), argv[0]);
+			return (1);
+		}
+	}
 
+	if ((fd = open(vdev, O_RDWR)) < 0) {
+		(void) fprintf(stderr, gettext("failed to open %s: %s\n"),
+		    vdev, strerror(errno));
 		return (1);
 	}
 
-	if (inuse) {
-		switch (state) {
-		default:
-		case POOL_STATE_ACTIVE:
-		case POOL_STATE_SPARE:
-		case POOL_STATE_L2CACHE:
-			(void) fprintf(stderr,
-			    gettext("labelclear operation failed.\n"
-			    "\tVdev %s is a member (%s), of pool \"%s\".\n"
-			    "\tTo remove label information from this device, "
-			    "export or destroy\n\tthe pool, or remove %s from "
-			    "the configuration of this pool\n\tand retry the "
-			    "labelclear operation.\n"),
-			    vdev, zpool_pool_state_to_name(state), name, vdev);
-			ret = 1;
-			goto errout;
+	if (zpool_read_label(fd, &config, NULL) != 0 || config == NULL) {
```

Comment:
It appears that the add-o_ashift test is failing consistently with the `failed to check state` error. The test case does a `zpool labelclear` on disk2 which didn't fail previously. Maybe change the test case to have a `log_mustnot` in the invalid ashift value for loop.